### PR TITLE
[AGENT] Harden transcription final pass and replay tooling

### PR DESCRIPTION
**Pull request by BASIC-BIT**
## Summary
- harden transcription voting and final-pass reconciliation against punctuation-only outputs, repeated low-information hallucinations, and segment-boundary drift
- add replay-oriented eval tooling for prompt vs no-prompt comparison, plus targeted tests and docs updates for the final-pass pipeline
- keep transcription observability high with extra trivial-text telemetry and config/docs coverage for the new safeguards

## Testing
- `yarn test --runTestsByPath src/utils/__tests__/transcriptionText.test.ts src/utils/__tests__/transcriptionVote.test.ts src/services/__tests__/transcriptionFinalPassService.test.ts`
- `yarn lint:check src/evals/transcriptionEval.ts src/services/transcriptionFinalPassService.ts src/services/transcriptionService.ts src/utils/transcriptionVote.ts src/utils/transcriptionText.ts src/utils/__tests__/transcriptionText.test.ts src/utils/__tests__/transcriptionVote.test.ts src/services/__tests__/transcriptionFinalPassService.test.ts`
- `yarn build`
- `npx prettier --check src/evals/transcriptionEval.ts src/services/transcriptionFinalPassService.ts src/services/transcriptionService.ts src/utils/transcriptionVote.ts src/utils/transcriptionText.ts src/utils/__tests__/transcriptionText.test.ts src/utils/__tests__/transcriptionVote.test.ts src/services/__tests__/transcriptionFinalPassService.test.ts docs/audio-transcription.md docs/feature-toggles.md docs/audio-transcription-plan.md`

## Notes
- `yarn markdownlint:check` is not runnable in this environment because `markdownlint-cli2` is not installed.

### CODE REVIEW

File: src/services/transcriptionFinalPassService.ts [hunk 1]
```diff
@@ -0,0 +1,960 @@
...
+  let repetitionFilteredSegments = 0;
+
+  for (const segment of baselineSegments) {
+    const text = resolveEffectiveSegmentText(segment).trim();
+    if (!isLowInformationTranscriptionText(text)) {
+      continue;
+    }
+
+    const speakerEntries = lastKeptBySpeaker.get(segment.speaker) ?? [];
+    const duplicate = speakerEntries.some((entry) => {
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/transcriptionFinalPassService.ts]: [AGENT] Switched the repeated low-information filter to key on the stable user ID instead of the display label, and added coverage for two users sharing the same nickname.

File: src/services/transcriptionService.ts [hunk 2]
```diff
@@ -359,12 +368,16 @@ async function transcribeInternal(
         transcriptSyllableCount: transcriptStats.transcriptSyllableCount,
         wordsPerSecond: transcriptStats.wordsPerSecond,
         syllablesPerSecond: transcriptStats.syllablesPerSecond,
+        rawTextQuality,
+        selectedTextQuality,
         transcriptionLength: guardResult.text.length,
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/transcriptionService.ts]: [AGENT] Updated the warning log payload to keep only counts, flags, and reasons. The raw transcript text is no longer included in those log objects.

File: src/evals/transcriptionEval.ts [hunk 3]
```diff
@@ -753,49 +802,64 @@ async function runFileEval(options: EvalOptions) {
...
+          });
+          const s3Client = new S3Client({ region });
+          return () =>
+            transcribeWithBedrock(
+              bedrockClient,
+              s3Client,
+              filePath,
+              bedrockConfig,
+            );
+        })();
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/evals/transcriptionEval.ts:842]: [AGENT] Restricted `--compare-no-prompt` to the OpenAI provider so the eval cannot run a misleading prompt-vs-no-prompt comparison on the Bedrock path.

File: test/commands/endMeeting.test.ts [hunk 4]
```diff
@@ -154,11 +158,27 @@ const mockedReleaseMeetingLeaseForMeeting =
...
+      applied: false,
+      processedChunks: 0,
+      totalChunks: 0,
+      totalSegments: 0,
+      candidateEdits: 0,
+      acceptedEdits: 0,
+      replacedSegments: 0,
+      droppedSegments: 0,
+      fallbackApplied: false,
+    });
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply test/commands/endMeeting.test.ts:183]: [AGENT] Added the missing `rejectedTrivialEdits` and `repetitionFilteredSegments` fields to the `runTranscriptionFinalPass` test mock in `test/commands/endMeeting.test.ts`.

File: src/utils/transcriptionText.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        break;
+      }
+    }
+    if (matched) {
+      return true;
+    }
+  }
+
+  return false;
+};
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionText.ts]: [AGENT] Renamed the helper to `isContiguousTokenSubarray` and added an inline note documenting the intentional phrase-containment behavior.

File: src/services/__tests__/transcriptionFinalPassService.test.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
+import { runTranscriptionFinalPass } from "../transcriptionFinalPassService";
+import type { MeetingData } from "../../types/meeting-data";
+
+const createMeeting = (transcript: string): MeetingData => {
+  const startTime = new Date("2026-03-01T00:00:00.000Z");
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/__tests__/transcriptionFinalPassService.test.ts:5]: [AGENT] Keeping both suites intentionally. `src/services/__tests__` covers the focused utility and repetition-filter edge cases close to the implementation, while `test/services` keeps the broader chunking and guardrail integration coverage in the existing service-level suite.

**Comment by BASIC-BIT** [comment]
[AGENT] Addressed the concrete bot review items locally before this push: added the missing final-pass mock fields in `test/commands/endMeeting.test.ts`, switched repeated-phrase tracking to stable user IDs, removed transcript text from the new warning log payloads, restricted `--compare-no-prompt` to OpenAI, and renamed the contiguous-token helper with an intent comment. I am intentionally keeping the focused `src/services/__tests__` coverage alongside the broader `test/services` suite.

File: src/utils/transcriptionText.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+const normalizeTranscriptWords = (text: string) =>
+  text
+    .toLowerCase()
+    .replace(/[^a-z0-9'\s]/g, " ")
+    .split(/\s+/)
+    .filter(Boolean);
+
+const countAlnumChars = (text: string) =>
+  Array.from(text).filter((char) => /[A-Za-z0-9]/.test(char)).length;
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionText.ts]: [AGENT] Switched the text-quality helpers to Unicode-aware letter and number classes so non-Latin transcripts are not misclassified as punctuation-only or trivial, and added a regression test for non-Latin input.

File: src/services/transcriptionFinalPassService.ts [hunk 8]
```diff
@@ -634,6 +671,54 @@ const applyAcceptedEdits = (
...
+    });
+
+    if (duplicate) {
+      segment.fileData.finalPassTranscript = "";
+      repetitionFilteredSegments += 1;
+      continue;
+    }
+
+    speakerEntries.push({ text, offsetSeconds: segment.offsetSeconds });
+    lastKeptBySpeaker.set(segment.speakerKey, speakerEntries);
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/transcriptionFinalPassService.ts]: [AGENT] Pruned each speaker queue to the active repeat window before duplicate checks, so the repeated-phrase filter no longer retains stale entries across long meetings.

File: src/utils/transcriptionVote.ts [hunk 9]
```diff
@@ -189,6 +204,15 @@ function chooseSelectedCandidateId(input: {
   promptQuality: CandidateQuality;
   noPromptQuality: CandidateQuality;
 }): TranscriptionVoteCandidateId {
+  if (input.noPromptQuality.trivialText) {
+    return "prompt";
+  }
+  if (input.promptQuality.trivialText && !input.noPromptQuality.trivialText) {
+    return "no_prompt";
+  }
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionVote.ts]: [AGENT] Dropped the redundant `!noPromptQuality.trivialText` clause from the prompt-trivial branch, since the earlier no-prompt trivial guard already covers that case.

File: src/utils/transcriptionVote.ts [hunk 10]
```diff
@@ -143,12 +152,18 @@ function calculateCandidateQuality(
   if (maxConsecutiveRepeats >= SCORE_CONSECUTIVE_REPEAT_MIN) {
     score += SCORE_CONSECUTIVE_REPEAT_PENALTY;
   }
+  if (textQuality.trivial) {
+    score += SCORE_TRIVIAL_TEXT_PENALTY;
+  }
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionVote.ts:159]: [AGENT] Added an inline note clarifying that `SCORE_TRIVIAL_TEXT_PENALTY` only affects score telemetry because trivial candidates short-circuit before score comparison.

File: src/services/transcriptionFinalPassService.ts [hunk 11]
```diff
@@ -439,17 +452,25 @@ const applyBatchEdits = (
...
   for (const edit of edits) {
     candidateEdits += 1;
     const segment = getSegmentById(batch, edit.segmentId);
+    if (edit.action === "replace") {
+      const replacement = (edit.text ?? "").trim();
+      if (replacement && isTrivialTranscriptionText(replacement)) {
+        rejectedTrivialEdits += 1;
+        continue;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/transcriptionFinalPassService.ts:470]: [AGENT] Moved the segment usability guard ahead of the trivial-text counter so `rejectedTrivialEdits` now reflects only edits that could otherwise have been applied.

File: src/evals/transcriptionEval.ts [hunk 12]
```diff
@@ -753,49 +808,64 @@ async function runFileEval(options: EvalOptions) {
...
+
+  const runBatchForPrompt = async (promptOverride: string) =>
+    await runTranscriptionBatch({
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/evals/transcriptionEval.ts]: [AGENT] Removed the redundant `async/await` wrapper around `runBatchForPrompt`, so it now returns the batch promise directly.

File: src/utils/transcriptionText.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  return getTranscriptionTextQuality(text).trivial;
+}
+
+export function isLowInformationTranscriptionText(text: string): boolean {
+  const quality = getTranscriptionTextQuality(text);
+  return (
+    !quality.trivial &&
+    quality.wordCount > 0 &&
+    quality.wordCount <= LOW_INFORMATION_MAX_WORDS
+  );
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionText.ts:69]: [AGENT] Tightened the low-information heuristic so multi-word phrases still use the word-count rule, while single-token text now has to stay under a separate alphanumeric-character cap. That keeps longer no-space transcripts from being treated as low-information.

File: src/utils/transcriptionText.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
+import { countWords } from "./text";
+
+const FAILURE_PLACEHOLDER = "[Transcription failed]";
+const LOW_INFORMATION_MAX_WORDS = 8;
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionText.ts]: [AGENT] Moved the failure sentinel into `constants.ts` and updated the transcription and audio paths to use the shared placeholder value.

File: src/utils/transcriptionVote.ts [hunk 15]
```diff
@@ -210,6 +236,19 @@ export function decideTranscriptionVote(input: {
...
+  if (promptQuality.trivialText !== noPromptQuality.trivialText) {
+    reasons.push("trivial_text_difference");
+  }
+  if (selectedId === "prompt" && noPromptQuality.trivialText) {
+    reasons.push("no_prompt_trivial_text");
+  }
+  if (
+    selectedId === "prompt" &&
+    !noPromptQuality.trivialText &&
+    noPromptQuality.score < NO_PROMPT_MIN_SELECTION_SCORE
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionVote.ts:254]: [AGENT] Added the symmetric trivial-text reason codes so vote telemetry now distinguishes `prompt_trivial_text`, `no_prompt_trivial_text`, and `both_trivial_text` cases during replay and log review.

File: src/utils/__tests__/transcriptionText.test.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    expect(quality.punctuationOnly).toBe(false);
+    expect(quality.alnumCharCount).toBeGreaterThan(0);
+  });
+
+  it("does not treat longer no-space text as low information", () => {
+    expect(isLowInformationTranscriptionText("こんにちは世界")).toBe(true);
+    expect(
+      isLowInformationTranscriptionText("これはかなり長めの日本語テキストです"),
+    ).toBe(false);
+  });
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/__tests__/transcriptionText.test.ts]: [AGENT] Renamed the test so it matches both assertions: short no-space text can still be low-information, while longer no-space text is not.

**Comment by BASIC-BIT** [comment]
[AGENT] Addressed the last minor cleanup items from the latest Greptile pass: the no-space text test now has an accurate description, and the guardrail-threshold fallback in `transcriptionFinalPassService` now documents why it intentionally skips the deterministic repetition filter when the reconcile model looks suspicious.

File: src/utils/transcriptionVote.ts [hunk 17]
```diff
@@ -95,8 +100,9 @@ export function hasLowConfidenceLogprobs(metrics?: LogprobMetrics): boolean {
 function calculateCandidateQuality(
   candidate: TranscriptionVoteCandidate,
 ): CandidateQuality {
-  const trimmed = candidate.text.trim();
-  const wordCount = countWords(trimmed);
+  const textQuality = getTranscriptionTextQuality(candidate.text);
+  const trimmed = textQuality.trimmed;
+  const wordCount = textQuality.wordCount;
   const normalizedWords = normalizeWords(trimmed);
   const uniqueWordRatio =
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/transcriptionVote.ts:107]: [AGENT] Updated `normalizeWords()` to use the same Unicode-aware letter and number classes as the text-quality helper, so uniqueness and repetition scoring stay aligned for non-ASCII transcripts.

File: src/utils/__tests__/transcriptionText.test.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    expect(quality.trivial).toBe(false);
+    expect(quality.punctuationOnly).toBe(false);
+    expect(quality.alnumCharCount).toBeGreaterThan(0);
+  });
+
+  it("does not treat longer no-space text as low information", () => {
+    expect(isLowInformationTranscriptionText("こんにちは世界")).toBe(true);
+    expect(
+      isLowInformationTranscriptionText("これはかなり長めの日本語テキストです"),
+    ).toBe(false);
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/utils/__tests__/transcriptionText.test.ts]: [AGENT] The test name is already updated locally to describe both assertions: short no-space text can be low-information, while longer no-space text is not.

File: src/services/transcriptionFinalPassService.ts [hunk 19]
```diff
@@ -634,6 +671,51 @@ const applyAcceptedEdits = (
...
+      continue;
+    }
+
+    recentEntries.push({ text, offsetSeconds: segment.offsetSeconds });
+    lastKeptBySpeaker.set(segment.speakerKey, recentEntries);
+  }
+
+  return repetitionFilteredSegments;
+};
 const defaultDependencies: FinalPassDependencies = {
```
[the remark this answers is not in the document]
Comment (BASIC-BIT) [reply src/services/transcriptionFinalPassService.ts:720]: [AGENT] Added the blank line so the repeated-filter helper and `defaultDependencies` declaration are visually separated again.
